File: combo2-advanced-agentic-engineering/tiny-agent-csharp/src/TinyAgent.Reference/Agent.cs
using System.Text;
using System.Text.Json.Nodes;
using TinyAgent.Shared;

namespace TinyAgent.Reference;

/// <summary>
/// The agent loop — complete worked solution.
/// </summary>
/// <remarks>
/// <para>Thesis (Thorsten Ball, ampcode.com): <i>It's an LLM, a loop, and enough tokens.</i></para>
/// <para>What the loop does, in one glance:</para>
/// <code>
/// contents = [user_prompt]
/// while turn &lt; maxTurns:
///     response = gemini.generate(contents, tools)
///     contents.Add(response.content)          // the most-forgotten line
///     calls = function calls in the response
///     if no calls: return the text            // done
///     foreach call: contents.Add(functionResponse)
/// </code>
/// <para>
/// Note it is <c>async</c>. The Python original is deliberately synchronous —
/// its facilitator notes say "not a chance to teach asyncio". In .NET every
/// HTTP call is a <see cref="Task"/>, so <c>await</c> is unavoidable here.
/// It is plumbing, not the lesson: read past it and look at the loop.
/// </para>
/// </remarks>
public static class Agent
{
    public const string SystemInstruction = """
        You are a careful coding assistant working inside a small
        code repository. You have three tools: read_file, list_files, and edit_file.

        Workflow:
        1. Explore first. Use list_files and read_file to build an understanding before editing.
        2. Edit sparingly. One edit per logical change. Use enough surrounding context in
           old_str so it matches exactly once.
        3. Report what you did in plain prose when you are finished. Do not call any tool on
           the final turn — that's how you signal you're done.
        4. If a tool returns a string starting with "ERROR:", read the error carefully and
           adjust your approach. Don't retry the same call blindly.
        """;

    public static async Task<string> RunAsync(
        string userPrompt,
        ITools tools,
        GeminiClient client,
        string? model = null,
        int maxTurns = 20,
        Action<AgentEvent>? onEvent = null,
        CancellationToken cancellationToken = default)
    {
        model ??= DotEnv.Model();

        // Conversation state. Gemini's "contents" is an ordered list of turns
        // alternating between role "user" and role "model". Tool results go back
        // as a *user* turn whose parts are functionResponse parts.
        var contents = new List<Content>
        {
            new("user", new[] { new Part(Text: userPrompt) }),
        };

        var declarations = new[] { ToolSchemas.All() };
        var systemInstruction = new Content("user", new[] { new Part(Text: SystemInstruction) });

        for (var turn = 1; turn <= maxTurns; turn++)
        {
            onEvent?.Invoke(new AgentEvent.TurnStart(turn));

            var request = new GeminiRequest(contents, declarations, systemInstruction);
            var response = await client.GenerateContentAsync(model, request, cancellationToken)
                                      .ConfigureAwait(false);

            var candidate = response.Candidates?.FirstOrDefault();
            if (candidate?.Content is null)
            {
                return $"ERROR: model returned no content (finishReason: {candidate?.FinishReason ?? "none"})";
            }

            // Append the model's turn BEFORE doing anything else. Forget this and
            // the model re-reads a context that never contains its own tool calls,
            // so it asks for the same thing forever. It is the #1 failure here.
            contents.Add(candidate.Content);

            var calls = (candidate.Content.Parts ?? Array.Empty<Part>())
                .Where(p => p.FunctionCall is not null)
                .Select(p => p.FunctionCall!)
                .ToList();

            if (calls.Count == 0)
            {
                // No tool calls -> the model signalled it is done.
                var finalText = string.Concat(
                    (candidate.Content.Parts ?? Array.Empty<Part>()).Select(p => p.Text ?? string.Empty));
                onEvent?.Invoke(new AgentEvent.Final(finalText, turn));
                return finalText;
            }

            // Execute every call and collect the responses.
            var responseParts = new List<Part>();
            foreach (var call in calls)
            {
                onEvent?.Invoke(new AgentEvent.ToolCall(call.Name, PreviewArgs(call.Args)));

                var result = Dispatch(tools, call);

                onEvent?.Invoke(new AgentEvent.ToolResult(call.Name, result));
                responseParts.Add(new Part(FunctionResponse: new FunctionResponse(
                    call.Name,
                    new JsonObject { ["result"] = result })));
            }

            // Send all tool responses back in a single user turn.
            contents.Add(new Content("user", responseParts));
        }

        return $"ERROR: agent did not finish within {maxTurns} turns";
    }

    /// <summary>
    /// Route one function call to a tool. Every failure path returns a string —
    /// nothing thrown here reaches the loop.
    /// </summary>
    public static string Dispatch(ITools tools, FunctionCall call)
    {
        try
        {
            return call.Name switch
            {
                ToolSchemas.ReadFileName => tools.ReadFile(RequiredArg(call, "path")),
                ToolSchemas.ListFilesName => tools.ListFiles(OptionalArg(call, "path") ?? ".").ToModelString(),
                ToolSchemas.EditFileName => tools.EditFile(
                    RequiredArg(call, "path"),
                    RequiredArg(call, "old_str"),
                    RequiredArg(call, "new_str")),
                _ => $"ERROR: unknown tool '{call.Name}'",
            };
        }
        catch (ArgumentException ex)
        {
            return $"ERROR: bad arguments to {call.Name}: {ex.Message}";
        }
        catch (Exception ex)
        {
            // Surface any tool failure to the model rather than killing the loop.
            return $"ERROR: {ex.GetType().Name}: {ex.Message}";
        }
    }

    private static string RequiredArg(FunctionCall call, string name) =>
        OptionalArg(call, name) ?? throw new ArgumentException($"missing required argument '{name}'");

    private static string? OptionalArg(FunctionCall call, string name)
    {
        if (call.Args is null) return null;
        return call.Args.TryGetPropertyValue(name, out var node) && node is not null
            ? node.GetValue<string>()
            : null;
    }

    private static string PreviewArgs(JsonObject? args)
    {
        if (args is null || args.Count == 0) return string.Empty;

        var sb = new StringBuilder();
        foreach (var (key, value) in args)
        {
            if (sb.Length > 0) sb.Append(", ");
            sb.Append(key).Append('=').Append(value?.ToJsonString() ?? "null");
        }

        var preview = sb.ToString();
        return preview.Length > 120 ? preview[..117] + "..." : preview;
    }
}

File: combo2-advanced-agentic-engineering/java/tiny-agent-java/reference/src/main/java/ai/octoco/tinyagent/reference/Agent.java
package ai.octoco.tinyagent.reference;

import ai.octoco.tinyagent.shared.AgentEvent;
import ai.octoco.tinyagent.shared.Content;
import ai.octoco.tinyagent.shared.DotEnv;
import ai.octoco.tinyagent.shared.FunctionCall;
import ai.octoco.tinyagent.shared.FunctionResponse;
import ai.octoco.tinyagent.shared.GeminiClient;
import ai.octoco.tinyagent.shared.GeminiRequest;
import ai.octoco.tinyagent.shared.Part;
import ai.octoco.tinyagent.shared.ToolSchemas;
import ai.octoco.tinyagent.shared.Tools;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.function.Consumer;

/**
 * The agent loop — complete worked solution.
 *
 * <p>Thesis (Thorsten Ball, ampcode.com): <i>It's an LLM, a loop, and enough
 * tokens.</i> What the loop does, in one glance:
 *
 * <pre>{@code
 * contents = [user_prompt]
 * while turn < maxTurns:
 *     response = gemini.generate(contents, tools)
 *     contents.add(response.content)          // the most-forgotten line
 *     calls = function calls in the response
 *     if no calls: return the text            // done
 *     foreach call: contents.add(functionResponse)
 * }</pre>
 *
 * <p>Note {@link #runAsync} returns a {@link CompletableFuture}. The Python
 * original is deliberately synchronous — its facilitator notes say "not a
 * chance to teach asyncio". In Java, {@code HttpClient.sendAsync} returns a
 * {@code CompletableFuture}, so each turn calls {@code .join()} on it —
 * the Java equivalent of {@code await}. It is plumbing, not the lesson: read
 * past it and look at the loop.
 */
public final class Agent {

    public static final String SYSTEM_INSTRUCTION = """
            You are a careful coding assistant working inside a small
            code repository. You have three tools: read_file, list_files, and edit_file.

            Workflow:
            1. Explore first. Use list_files and read_file to build an understanding before editing.
            2. Edit sparingly. One edit per logical change. Use enough surrounding context in
               old_str so it matches exactly once.
            3. Report what you did in plain prose when you are finished. Do not call any tool on
               the final turn — that's how you signal you're done.
            4. If a tool returns a string starting with "ERROR:", read the error carefully and
               adjust your approach. Don't retry the same call blindly.
            """;

    private Agent() {}

    public static CompletableFuture<String> runAsync(String userPrompt, Tools tools, GeminiClient client) {
        return runAsync(userPrompt, tools, client, null, 20, evt -> { });
    }

    public static CompletableFuture<String> runAsync(
            String userPrompt, Tools tools, GeminiClient client, String model, int maxTurns) {
        return runAsync(userPrompt, tools, client, model, maxTurns, evt -> { });
    }

    public static CompletableFuture<String> runAsync(
            String userPrompt,
            Tools tools,
            GeminiClient client,
            String model,
            int maxTurns,
            Consumer<AgentEvent> onEvent) {
        String resolvedModel = model != null ? model : DotEnv.model();

        // Conversation state. Gemini's "contents" is an ordered list of turns
        // alternating between role "user" and role "model". Tool results go back
        // as a *user* turn whose parts are functionResponse parts.
        List<Content> contents = new ArrayList<>();
        contents.add(new Content("user", List.of(Part.text(userPrompt))));

        List<JsonNode> declarations = List.of(ToolSchemas.all());
        Content systemInstruction = new Content("user", List.of(Part.text(SYSTEM_INSTRUCTION)));

        return CompletableFuture.supplyAsync(() -> {
            for (int turn = 1; turn <= maxTurns; turn++) {
                onEvent.accept(new AgentEvent.TurnStart(turn));

                GeminiRequest request = new GeminiRequest(contents, declarations, systemInstruction);
                var response = client.generateContentAsync(resolvedModel, request).join();

                var candidate = response.firstCandidate();
                if (candidate == null || candidate.content() == null) {
                    String finishReason = candidate != null ? candidate.finishReason() : "none";
                    return "ERROR: model returned no content (finishReason: "
                            + (finishReason != null ? finishReason : "none") + ")";
                }

                // Append the model's turn BEFORE doing anything else. Forget this and
                // the model re-reads a context that never contains its own tool calls,
                // so it asks for the same thing forever. It is the #1 failure here.
                contents.add(candidate.content());

                List<FunctionCall> calls = new ArrayList<>();
                List<Part> parts = candidate.content().parts();
                if (parts != null) {
                    for (Part part : parts) {
                        if (part.functionCall() != null) {
                            calls.add(part.functionCall());
                        }
                    }
                }

                if (calls.isEmpty()) {
                    // No tool calls -> the model signalled it is done.
                    StringBuilder finalText = new StringBuilder();
                    if (parts != null) {
                        for (Part part : parts) {
                            if (part.text() != null) finalText.append(part.text());
                        }
                    }
                    onEvent.accept(new AgentEvent.Final(finalText.toString(), turn));
                    return finalText.toString();
                }

                // Execute every call and collect the responses.
                List<Part> responseParts = new ArrayList<>();
                for (FunctionCall call : calls) {
                    onEvent.accept(new AgentEvent.ToolCall(call.name(), previewArgs(call.args())));

                    String result = dispatch(tools, call);

                    onEvent.accept(new AgentEvent.ToolResult(call.name(), result));
                    ObjectNode resultNode = JsonNodeFactory.instance.objectNode().put("result", result);
                    responseParts.add(Part.functionResponse(new FunctionResponse(call.name(), resultNode)));
                }

                // Send all tool responses back in a single user turn.
                contents.add(new Content("user", responseParts));
            }

            return "ERROR: agent did not finish within " + maxTurns + " turns";
        });
    }

    /**
     * Route one function call to a tool. Every failure path returns a string —
     * nothing thrown here reaches the loop.
     */
    public static String dispatch(Tools tools, FunctionCall call) {
        try {
            return switch (call.name()) {
                case ToolSchemas.READ_FILE -> tools.readFile(requiredArg(call, "path"));
                case ToolSchemas.LIST_FILES -> {
                    String path = optionalArg(call, "path");
                    yield tools.listFiles(path != null ? path : ".").toModelString();
                }
                case ToolSchemas.EDIT_FILE -> tools.editFile(
                        requiredArg(call, "path"),
                        requiredArg(call, "old_str"),
                        requiredArg(call, "new_str"));
                default -> "ERROR: unknown tool '" + call.name() + "'";
            };
        } catch (IllegalArgumentException ex) {
            return "ERROR: bad arguments to " + call.name() + ": " + ex.getMessage();
        } catch (Exception ex) {
            // Surface any tool failure to the model rather than killing the loop.
            return "ERROR: " + ex.getClass().getSimpleName() + ": " + ex.getMessage();
        }
    }

    private static String requiredArg(FunctionCall call, String name) {
        String value = optionalArg(call, name);
        if (value == null) throw new IllegalArgumentException("missing required argument '" + name + "'");
        return value;
    }

    private static String optionalArg(FunctionCall call, String name) {
        JsonNode args = call.args();
        if (args == null) return null;
        JsonNode node = args.get(name);
        return (node == null || node.isNull()) ? null : node.asText();
    }

    private static String previewArgs(JsonNode args) {
        if (args == null || args.isEmpty()) return "";

        StringBuilder sb = new StringBuilder();
        var fields = args.fields();
        while (fields.hasNext()) {
            var entry = fields.next();
            if (sb.length() > 0) sb.append(", ");
            sb.append(entry.getKey()).append('=').append(entry.getValue().toString());
        }

        String preview = sb.toString();
        return preview.length() > 120 ? preview.substring(0, 117) + "..." : preview;
    }
}
